api/py/ai/chronon/repo/validator.py

```diff
@@ -87,52 +103,87 @@ def get_group_by_output_columns(group_by: GroupBy) -> List[str]:
     """
     output_columns = set(get_pre_derived_group_by_columns(group_by))
-
     if group_by.derivations:
-        # if derivations contain star, then all columns are included except the columns which are renamed
-        found = any(derivation.expression == "*" for derivation in group_by.derivations)
-        if not found:
-            output_columns.clear()
-        for derivation in group_by.derivations:
-            if found and is_identifier(derivation.expression):
-                output_columns.remove(derivation.expression)
-            if derivation.name != "*":
-                output_columns.add(derivation.name)
-
-    return list(output_columns)
+        return build_derived_columns(output_columns, group_by.derivations)
+    else:
+        return list(output_columns)
 
 
-def get_pre_derived_join_columns(join: Join) -> List[str]:
-    output_columns = []
+def get_pre_derived_join_features(join: Join) -> List[str]:
+    internal_features = []
     for jp in join.joinParts:
-        group_by_cols = get_group_by_output_columns(jp.groupBy)
-        for col in group_by_cols:
+        pre_derived_group_by_features = set(get_pre_derived_group_by_features(jp.groupBy))
+        derived_group_by_features = build_derived_columns(pre_derived_group_by_features, jp.groupBy.derivations)
+        for col in derived_group_by_features:
             prefix = jp.prefix + "_" if jp.prefix else ""
             gb_prefix = jp.groupBy.metaData.name.replace(".", "_")
-            output_columns.append(prefix + gb_prefix + "_" + col)
-    return output_columns
+            internal_features.append(prefix + gb_prefix + "_" + col)
+    return internal_features
 
 
-def get_join_output_columns(join: Join) -> List[str]:
+def get_pre_derived_join_columns(join: Join) -> List[str]:
+    internal_columns = []
+    if join.left.events:
+        internal_columns.extend(join.left.events.query.selects.keys())
+    elif join.left.entities:
+        internal_columns.extend(join.left.entities.query.selects.keys())
+    elif join.left.joinSource:
+        internal_columns.extend(join.left.joinSource.query.selects.keys())
+    internal_columns.extend(get_pre_derived_join_features(join))
+    return internal_columns + get_external_columns(join)
+
+
+# The logic should be consistent with the full name logic defined
+# in https://github.com/airbnb/chronon/blob/main/api/src/main/scala/ai/chronon/api/Extensions.scala#L677.
+def get_external_part_full_name(external_part: ExternalPart) -> str:
+    prefix = external_part.prefix + "_" if external_part.prefix else ""
+    name = external_part.source.metadata.name
+    sanitized_name = re.sub("[^a-zA-Z0-9_]", "_", name)
+    return "ext_" + prefix + sanitized_name
+
+
+# The external columns name logic should be consistent with the logic defined in fetcher.scala
+# https://github.com/airbnb/chronon/blob/main/online/src/main/scala/ai/chronon/online/Fetcher.scala#L371
+def get_external_columns(join: Join) -> List[str]:
```

Comment:
nit: get_external_features

Reply:
One question here is, external part is purely an online concept right? And since we don't include external part keys for online case does it mean external_features = external_columns?

Reply:
updated